File: Hassil.Api/Services/Invoices/InvoiceService.cs
using Hassil.Api.Contracts.Invoices;
using Hassil.Api.Database;
using Hassil.Api.Domain.Enums;
using Hassil.Api.Domain.Models;
using Hassil.Api.Exceptions;
using Microsoft.EntityFrameworkCore;
using ValidationException = Hassil.Api.Exceptions.ValidationException;

namespace Hassil.Api.Services.Invoices;

public class InvoiceService(
    HassilDbContext dbContext,
    IInvoiceFingerprintService fingerprintService,
    ILogger<InvoiceService> logger) : IInvoiceService
{
    public async Task<Invoice> CreateAsync(
        Guid userId,
        CreateInvoiceRequest request,
        ReceivableSource receivableSource,
        CancellationToken ct = default)
    {
        EnsureRequired(request.ClientName, "Client name is required.", "CLIENT_NAME_REQUIRED");
        EnsureRequired(request.ClientEmail, "Client email is required.", "CLIENT_EMAIL_REQUIRED");
        EnsureRequired(request.InvoiceNumber, "Invoice number is required.", "INVOICE_NUMBER_REQUIRED");
        EnsureRequired(request.Currency, "Currency is required.", "CURRENCY_REQUIRED");

        if (request.Amount <= 0)
            throw new ValidationException(
                "Invoice amount must be greater than zero.",
                "INVALID_INVOICE_AMOUNT");

        if (request.DueDate < request.IssueDate)
            throw new ValidationException(
                "Due date cannot be before issue date.",
                "INVALID_DUE_DATE");

        var client = await GetOrCreateClientAsync(
            request.ClientName,
            request.ClientEmail,
            request.ClientCountry,
            ct);

        var fingerprint = fingerprintService.Create(
            request.InvoiceNumber,
            request.ClientEmail,
            request.Amount,
            request.DueDate,
            receivableSource);

        var duplicateExists = await dbContext.Invoices
            .AnyAsync(i => i.InvoiceFingerprint == fingerprint, ct);

        if (duplicateExists)
            throw new ConflictException(
                "An invoice with the same number, client, amount, due date, and source already exists.",
                "DUPLICATE_INVOICE");

        var invoice = Invoice.Create(
            userId:             userId,
            clientId:           client.Id,
            invoiceNumber:      request.InvoiceNumber,
            amount:             request.Amount,
            issueDate:          request.IssueDate,
            dueDate:            request.DueDate,
            invoiceFingerprint: fingerprint,
            receivableSource:   receivableSource,
            currency:           request.Currency,
            description:        request.Description,
            paymentTerms:       request.PaymentTerms);

        dbContext.Invoices.Add(invoice);
        await dbContext.SaveChangesAsync(ct);

        logger.LogInformation(
            "Invoice created. InvoiceId={InvoiceId} UserId={UserId} InvoiceNumber={InvoiceNumber}",
            invoice.Id,
            userId,
            invoice.InvoiceNumber);

        await LoadInvoiceGraphAsync(invoice, ct);
        return invoice;
    }

    public async Task<IReadOnlyList<Invoice>> GetInvoicesAsync(
        Guid userId,
        CancellationToken ct = default)
    {
        return await dbContext.Invoices
            .Where(i => i.UserId == userId)
            .Include(i => i.Client)
            .Include(i => i.Documents)
            .Include(i => i.AdvanceRequest)
            .OrderByDescending(i => i.CreatedAt)
            .ToListAsync(ct);
    }

    public async Task<Invoice> GetInvoiceAsync(
        Guid userId,
        Guid invoiceId,
        CancellationToken ct = default)
    {
        return await InvoiceQuery(userId)
                   .FirstOrDefaultAsync(i => i.Id == invoiceId, ct)
               ?? throw new NotFoundException("Invoice not found.", "INVOICE_NOT_FOUND");
    }

    public async Task<InvoiceDocument> AddDocumentAsync(
        Guid userId,
        Guid invoiceId,
        AddInvoiceDocumentRequest request,
        CancellationToken ct = default)
    {
        EnsureRequired(request.FileName, "File name is required.", "FILE_NAME_REQUIRED");
        EnsureRequired(request.DocumentType, "Document type is required.", "DOCUMENT_TYPE_REQUIRED");

        var invoice = await GetInvoiceAsync(userId, invoiceId, ct);

        if (invoice.Status is InvoiceStatus.Cancelled or InvoiceStatus.Paid)
            throw new ConflictException(
                $"Invoice '{invoice.InvoiceNumber}' cannot accept documents from status '{invoice.Status}'.",
                "INVALID_INVOICE_STATUS");

        var fileUrl = string.IsNullOrWhiteSpace(request.FileUrl)
            ? $"/demo-documents/{Uri.EscapeDataString(request.FileName.Trim())}"
            : request.FileUrl.Trim();

        var document = InvoiceDocument.Create(
            invoiceId:     invoice.Id,
            fileName:      request.FileName,
            fileUrl:       fileUrl,
            documentType:  request.DocumentType);

        invoice.AddDocument(document);
        await dbContext.SaveChangesAsync(ct);

        logger.LogInformation(
            "Invoice document added. InvoiceId={InvoiceId} DocumentId={DocumentId}",
            invoice.Id,
            document.Id);

        return document;
    }

    public async Task<Invoice> SubmitAsync(
        Guid userId,
        Guid invoiceId,
        CancellationToken ct = default)
    {
        var invoice = await GetInvoiceAsync(userId, invoiceId, ct);

        try
        {
            invoice.Submit();
        }
        catch (InvalidOperationException ex)
        {
            throw new ConflictException(ex.Message, "INVALID_INVOICE_TRANSITION");
        }

        await dbContext.SaveChangesAsync(ct);

        logger.LogInformation(
            "Invoice submitted. InvoiceId={InvoiceId} UserId={UserId}",
            invoice.Id,
            userId);

        return invoice;
    }

    private IQueryable<Invoice> InvoiceQuery(Guid userId) =>
        dbContext.Invoices
            .Where(i => i.UserId == userId)
            .Include(i => i.Client)
            .Include(i => i.Documents)
            .Include(i => i.AdvanceRequest);

    private async Task<Client> GetOrCreateClientAsync(
        string name,
        string email,
        string? country,
        CancellationToken ct)
    {
        var normalizedEmail = email.Trim().ToLowerInvariant();

        var client = await dbContext.Clients
            .FirstOrDefaultAsync(c => c.Email == normalizedEmail, ct);

        if (client is not null)
            return client;

        client = Client.Create(name, normalizedEmail, country);
        dbContext.Clients.Add(client);

        return client;
    }

    private async Task LoadInvoiceGraphAsync(Invoice invoice, CancellationToken ct)
    {
        if (!dbContext.Entry(invoice).Reference(i => i.Client).IsLoaded)
            await dbContext.Entry(invoice).Reference(i => i.Client).LoadAsync(ct);

        if (!dbContext.Entry(invoice).Collection(i => i.Documents).IsLoaded)
            await dbContext.Entry(invoice).Collection(i => i.Documents).LoadAsync(ct);

        if (!dbContext.Entry(invoice).Reference(i => i.AdvanceRequest).IsLoaded)
            await dbContext.Entry(invoice).Reference(i => i.AdvanceRequest).LoadAsync(ct);
    }

    private static void EnsureRequired(string? value, string message, string code)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ValidationException(message, code);
    }
}

File: hassil-web/src/services/invoiceService.ts
import { api } from './apiClient'
import type { Invoice, InvoiceDocument, ReceivableSource } from '../types'

export interface CreateInvoicePayload {
    clientName: string
    clientEmail: string
    clientCountry?: string
    invoiceNumber: string
    receivableSource: ReceivableSource
    amount: number
    currency: string
    issueDate: string
    dueDate: string
    description?: string
    paymentTerms?: string
}

export interface AddInvoiceDocumentPayload {
    fileName: string
    fileUrl?: string
    documentType: string
}

interface ApiClient {
    id: string
    name: string
    email: string
    country?: string
}

interface ApiInvoiceDocument {
    id: string
    fileName: string
    fileUrl?: string
    documentType: string
    uploadedAt: string
}

interface ApiInvoiceResponse {
    id: string
    userId: string
    client: ApiClient
    invoiceNumber: string
    receivableSource: ReceivableSource
    amount: number
    currency: string
    issueDate: string
    dueDate: string
    description?: string
    paymentTerms?: string
    status: Invoice['status']
    invoiceFingerprint: string
    documents: ApiInvoiceDocument[]
    advanceRequestId?: string | null
    createdAt: string
    updatedAt: string
}

interface ApiInvoiceSummaryResponse {
    id: string
    invoiceNumber: string
    client: ApiClient
    receivableSource: ReceivableSource
    amount: number
    currency: string
    dueDate: string
    status: Invoice['status']
    documentCount: number
    advanceRequestId?: string | null
    createdAt: string
}

function mapDocument(document: ApiInvoiceDocument, invoiceId: string): InvoiceDocument {
    return {
        id: document.id,
        invoiceId,
        fileName: document.fileName,
        fileUrl: document.fileUrl,
        documentType: document.documentType,
        uploadedAt: document.uploadedAt,
    }
}

function mapInvoice(response: ApiInvoiceResponse): Invoice {
    return {
        id: response.id,
        userId: response.userId,
        clientId: response.client.id,
        client: response.client,
        invoiceNumber: response.invoiceNumber,
        receivableSource: response.receivableSource,
        amount: response.amount,
        currency: response.currency,
        issueDate: response.issueDate,
        dueDate: response.dueDate,
        description: response.description,
        paymentTerms: response.paymentTerms,
        status: response.status,
        fingerprint: response.invoiceFingerprint,
        invoiceFingerprint: response.invoiceFingerprint,
        createdAt: response.createdAt,
        updatedAt: response.updatedAt,
        documents: response.documents.map((document) => mapDocument(document, response.id)),
        documentCount: response.documents.length,
        advanceRequestId: response.advanceRequestId ?? undefined,
    }
}

function mapInvoiceSummary(response: ApiInvoiceSummaryResponse): Invoice {
    return {
        id: response.id,
        userId: '',
        clientId: response.client.id,
        client: response.client,
        invoiceNumber: response.invoiceNumber,
        receivableSource: response.receivableSource,
        amount: response.amount,
        currency: response.currency,
        issueDate: response.createdAt.slice(0, 10),
        dueDate: response.dueDate,
        status: response.status,
        fingerprint: '',
        createdAt: response.createdAt,
        documents: [],
        documentCount: response.documentCount,
        advanceRequestId: response.advanceRequestId ?? undefined,
    }
}

export const InvoiceService = {
    list: async () => {
        const response = await api.get<ApiInvoiceSummaryResponse[]>('/invoices')
        return response.map(mapInvoiceSummary)
    },

    get: async (id: string) => {
        const response = await api.get<ApiInvoiceResponse>(`/invoices/${id}`)
        return mapInvoice(response)
    },

    create: async (invoice: CreateInvoicePayload) => {
        const response = await api.post<ApiInvoiceResponse>('/invoices', invoice)
        return mapInvoice(response)
    },

    addDocument: async (id: string, document: AddInvoiceDocumentPayload) => {
        const response = await api.post<ApiInvoiceDocument>(`/invoices/${id}/documents`, document)
        return mapDocument(response, id)
    },

    submit: async (id: string) => {
        const response = await api.post<ApiInvoiceResponse>(`/invoices/${id}/submit`)
        return mapInvoice(response)
    },
}
